### PR TITLE
[Merged by Bors] - feat: miscellaneous linear algebra lemmas

### PR DESCRIPTION
**Pull request by ocfnash**
---

Split out from https://github.com/leanprover-community/mathlib4/pull/7963

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/Submodule/Lattice.lean [hunk 1]
```diff
@@ -130,12 +130,18 @@ def botEquivPUnit : (⊥ : Submodule R M) ≃ₗ[R] PUnit.{v+1} where
...
+  rw [subsingleton_iff, Submodule.eq_bot_iff]
+  refine ⟨fun h x hx ↦ by simpa using h ⟨x, hx⟩ ⟨0, p.zero_mem⟩,
+    fun h ⟨x, hx⟩ ⟨y, hy⟩ ↦ by simp [h x hx, h y hy]⟩
+
+theorem eq_bot_of_subsingleton [Subsingleton p] : p = ⊥ :=
+  subsingleton_iff_eq_bot.mp inferInstance
 #align submodule.eq_bot_of_subsingleton Submodule.eq_bot_of_subsingleton
 
+theorem nontrivial_iff_ne_bot : Nontrivial p ↔ p ≠ ⊥ := by
+  rw [← not_iff_not, ne_eq, not_not, not_nontrivial_iff_subsingleton, subsingleton_iff_eq_bot]
```
Comment (eric-wieser) [inline Mathlib/Algebra/Module/Submodule/Lattice.lean:143]: nit: could maybe golf with `not_iff_comm`

Comment (ocfnash) [reply Mathlib/Algebra/Module/Submodule/Lattice.lean]: Thanks, this is a job for `iff_not_comm`!

File: Mathlib/LinearAlgebra/DFinsupp.lean [hunk 2]
```diff
@@ -337,20 +337,35 @@ theorem biSup_eq_range_dfinsupp_lsum (p : ι → Prop) [DecidablePred p] (S : ι
...
+See also `Submodule.mem_iSup_iff_exists_finsupp`. -/
 theorem mem_iSup_iff_exists_dfinsupp' (p : ι → Submodule R N) [∀ (i) (x : p i), Decidable (x ≠ 0)]
     (x : N) : x ∈ iSup p ↔ ∃ f : Π₀ i, p i, (f.sum fun i xi => ↑xi) = x := by
   rw [mem_iSup_iff_exists_dfinsupp]
   simp_rw [DFinsupp.lsum_apply_apply, DFinsupp.sumAddHom_apply,
     LinearMap.toAddMonoidHom_coe, coeSubtype]
 #align submodule.mem_supr_iff_exists_dfinsupp' Submodule.mem_iSup_iff_exists_dfinsupp'
 
+lemma mem_iSup_iff_exists_finsupp (p : ι → Submodule R N) (x : N) :
+    x ∈ iSup p ↔ ∃ (f : ι →₀ N), (∀ i, f i ∈ p i) ∧ (f.sum fun _i xi ↦ xi) = x := by
```
[the remark this answers is not in the document]
Comment (ocfnash) [reply Mathlib/LinearAlgebra/DFinsupp.lean:360]: Of course `mem_iSup_iff_exists_dfinsupp'` would suffice, but once I found destructing the dependent values it provides, I decided this was what I really wanted.

If nothing else, using `mem_iSup_iff_exists_finsupp` places less of a demand on the library to have API for dependent functions.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Comment by ocfnash** [comment]
GitHub was flaking and bors crashed so we need to try this again

bors merge

**Pull request closed without merging**